packages/dds/tree/docs/wip/schema-evolution.md

```diff
@@ -0,0 +1,583 @@
+The developer must do a **staged rollout** - an incremental set of application deployments that release client code only when it is safe to do so.
+
+1. Initially, the documents are using stored schema A and the application is written against view schema A.
+2. View schema B is written into the application client code.
+   The application's UI, etc. is written against it, but the existing code that (written against view schema A) continues to be maintained.
+   Thus the application logic becomes bifurcated according to the schema.
+   If the application loads a document with stored schema A, it views that document with view schema A.
+   If instead it loads a document with stored schema B, it views it with view schema B.
+   Both versions of the view/UI/application/etc. must continue to be maintained until after step 4.
+   This new version of the application is deployed and clients can begin updating to it.
+3. The developer waits until all (or most) clients have upgraded to this new application version.
+   The amount of clients that have moved to the new version is known as the **client saturation**.
+   It may be impossible to know for sure when 100% client saturation has been achieved, so a developer may estimate when the client saturation is sufficient - e.g. after 99% of online clients in the last 30 days had updated the new version.
+   After client saturation, the stored schema of the documents can now be upgraded to B.
+   It's important that this happens after client saturation of view schema B, or else there will be many clients that fail to open a document with stored schema B because they only know about view schema A.
+   A new version of the application is deployed which will upgrade the document to stored schema B when it first loads (or first writes to) a document using stored schema A.
+
+   If every client is guaranteed to have write permissions, then the developer can also remove view schema A and the code written against it in the same deployment.
+   This is because any client that encounters a document with stored schema A can upgrade it to stored schema B before interacting with it.
+   However, if the client is a read-only client, then it can't do this, and it needs to retain the code to read view schema A in case it opens a document with stored schema A.
+   This caveat is made more manageable by [adapters](#adapters), introduced later in this document.
+
+     > To minimize the time between client saturation and document upgrades, the ability to upgrade the document can be made part of the deployment in step 1 but "shipped dark" and remotely enabled in this step via a feature flight.
+
+4. Wait until all documents have upgraded to the new stored schema.
+   After this **document saturation** has been achieved, the application can do another deployment, removing view schema A and all of its related code, if it was not able to do so in step 3 (because some of its clients are read-only).
```

Comment:
This section assumes this is achievable. This is not the case in most scenarios. The expectation that most developers and users would have - a document that was not opened for 20 years can still be opened. Even if it's a single document.